### PR TITLE
Convert to AsciiDoc

### LINKED ISSUE #95: Convert to AsciiDoc
In light of #87 I think it'd be nice if we adopted a better format for the guide, namely AsciiDoc. It would allow us to have native admonitions, footnotes, sane nesting of list items, etc and to easily export the guide in various formats. That would also make it trivial to publish a simple site from the guide down the road if we want to.

It's trivial to convert to AsciiDoc with this tool https://github.com/asciidoctor/kramdown-asciidoc (created by the maintainer of AsciiDoctor himself). 

In general my longer term plan is to convert all guides to AsciiDoc and maybe even RuboCop's own documentation.

### PR DESCRIPTION
**Pull request by pirj**
Fixes #95

The conversion went quite well, there's no visual difference between the old and the new, except for increased vertical whitespace in lists, you can notice it by looking at ToC.

I'm not too happy about one thing. The recommendations were using bulleted list, and it seems that in AsciiDoc multi-line text in lists are done by adding a `+` (line break?) to the beginning of the line, which doesn't make it more readable in the editor.

Tried using sections, to add `§`, however GitHub seems to add it to the sections anyway, with an id auto-generated from the title itself. Tried all possible combinations of `:sectids:` / `:sectanchors:` / `:sectlinks:` to no avail.

Another option would be to insert a Unicode bullet symbol to beginning of recommendations, and added horizontal rules between recommendations. Added it to a couple of recommendations in the in the beginning of the guide. [Please take a look](https://github.com/rubocop-hq/rspec-style-guide/blob/convert-to-asciidoc/README.adoc#layout) and let me know what you think.

### CODE REVIEW

**Comment by bbatsov** [comment]
> I'm not too happy about one thing. The recommendations were using bulleted list, and it seems that in AsciiDoc multi-line text in lists are done by adding a + (line break?) to the beginning of the line, which doesn't make it more readable in the editor.

Yeah, just saw what you mean here https://github.com/powerman/asciidoc-cheatsheet/blob/master/README.adoc Seems like a GitHub issue though, but it's definitely lame. 

> Another option would be to insert a Unicode bullet symbol to beginning of recommendations, and added horizontal rules between recommendations. Added it to a couple of recommendations in the in the beginning of the guide. Please take a look and let me know what you think.

That's fine, but it's a bit annoying to type. I wonder if there's a way to circumvent GitHub's default behaviour somehow. @mojavelinux might know some tricks, so I'll ping him.

**Comment by bbatsov** [comment]
Btw, here are few useful tips about AsciiDoc on GitHub https://gist.github.com/dcode/0cfbf2699a1fe9b46ff04c41721dda74

**Comment by bbatsov** [comment]
Have you seen this issue https://github.com/asciidoctor/asciidoctor/issues/1087 ? Seems exactly like the problem we're trying to solve here.

**Comment by mojavelinux** [comment]
> The recommendations were using bulleted list, and it seems that in AsciiDoc multi-line text in lists are done by adding a + (line break?) to the beginning of the line, which doesn't make it more readable in the editor.

The `+` is a list continuation. (See https://asciidoctor.org/docs/user-manual/#list-continuation) It allows you to attach additional blocks to a list item. It has nothing to do with multi-line text. You can have as many lines in the text of a list item as you want. It's about multiple blocks. If you want to attach a separate block, such as a paragraph or code block, then you need a list continuation. And this is not specific to GitHub. It's how AsciiDoc has always been (dating back almost 20 years).

If you are going to attach more than one block to a list item, I recommend enclosing the content in an open block, which saves you from having to use more than one list continuation per list item. In other words:

```
* item text
+
--
paragraph

[,ruby]
----
puts 1
----

another paragraph
--
```

**Comment by mojavelinux** [comment]
> Another option would be to insert a Unicode bullet symbol to beginning of recommendations, and added horizontal rules between recommendations.

I strongly discourage sacrificing semantics for presentation. That goes against what AsciiDoc is about. If you don't like having to attach blocks to a list item, then I would recommend switching to sections. The benefit is that the section title (aka heading) is linkable.

File: README.adoc [hunk 1]
```diff
@@ -0,0 +1,1763 @@
+= The RSpec Style Guide
+:doctype: article
```
Comment (mojavelinux) [inline README.adoc]: Not needed. article is the default doctype.

File: README.adoc [hunk 2]
```diff
@@ -0,0 +1,1763 @@
...
+
+This RSpec style guide outlines the recommended best practices for real-world
```
Comment (mojavelinux) [inline README.adoc]: I strong recommend not hard-wrapping paragraphs. That makes collaborative editing much more difficult.

File: README.adoc [hunk 3]
```diff
@@ -0,0 +1,1763 @@
...
+unclear, feel free to open an issue asking for further clarity.
+
+== A Living Document
+
+Per the comment above, this guide is a work in progress - some rules are simply
+lacking thorough examples, but some things in the RSpec world change week by
+week or month by month. With that said, as the standard changes this guide is
+meant to be able to change with it.
+
+== Table of Contents
```
Comment (mojavelinux) [inline README.adoc]: You don't need to hardcode the table of contents. This feature is already provided by AsciiDoc. Add `:toc: macro` to the document header, then use `toc::[]` to place the table of contents.

File: README.adoc [hunk 4]
```diff
@@ -0,0 +1,1763 @@
...
+----
+
+'''
+
+[[empty-lines-between-describes]]
+&bull; Leave one empty line between `feature`, `context` or `describe` blocks.
+Do not leave empty line after the last such block in a group.
+^[<<empty-lines-between-describes,link>>]^
+
+[source,ruby]
```
Comment (mojavelinux) [inline README.adoc]: The word "source" is no longer required. You can write this line as `[,ruby]`.

File: README.adoc [hunk 5]
```diff
@@ -0,0 +1,1763 @@
...
+
+# good
+let(:name) { 'John Wayne' }
+
+it 'reverses a name' do
+  expect(reverser.reverse(name)).to eq('enyaW nhoJ')
+end
+----
+
+* +++<a name="shared-examples">++++++</a>+++
```
Comment (mojavelinux) [inline README.adoc]: There's no need to use raw HTML to make an anchor on a list. You can write it as follows:

```
* [[shared-examples]]Use shared examples to reduce duplication.
```

**Comment by mojavelinux** [comment]
Using passthroughs (triple plus) is very bad practice. In all cases, they are being used to define an anchor, which is already supported in AsciiDoc (e.g., `[[anchor]]`).

**Comment by mojavelinux** [comment]
> Have you seen this issue asciidoctor/asciidoctor#1087 ? Seems exactly like the problem we're trying to solve here.

GitHub could add styles to their stylesheet to reduce the space between list items. However, thus far they have not showed a willingness to do so. The proposal for Asciidoctor is to change the default HTML converter to wrap the primary text of list items using a "span" instead of a "p". That would make the problem go away.

File: README.adoc (hunk 2 above)
Comment (pirj) [reply README.adoc]: Can you please elaborate on this?

I can imagine that at several concurrent changes to the same paragraph scale the merge would become painful. But won't it be mostly manual if it was represented as a single line?

File: README.adoc [hunk 6]
```diff
@@ -0,0 +1,1763 @@
...
+ ** <<views,Views>>
+ ** <<controllers,Controllers>>
+ ** <<models,Models>>
+ ** <<mailers,Mailers>>
+* <<recommendations,Recommendations>>
+
+== Layout
+
+[[empty-lines-after-describe]]
+&bull; Do not leave empty lines after `feature`, `context` or `describe` descriptions.
```
Comment (mojavelinux) [inline README.adoc]: Ugh. Using `&bull;` to mark a bulleted list item severally breaks semantics. I discourage this approach.

**Comment by mojavelinux** [comment]
> May I ask you for a tip on how to structure this document using sections?

You'd need to assign each rule a name. But since you already have IDs for each rule, you can simply reverse engineer a heading that would produce that ID. Then, you knock out a whole bunch of problems at once (backwards compatible IDs, floating anchors, automatic TOC, no need for list continuations). And it's just going to make the whole document easier to navigate.

Here's an example:

```
=== Empty Lines After Describe

Do not leave empty lines after `feature`, `context` or `describe` descriptions.
It doesn't make the code more readable and lowers the value of logical chunks.
```

> that diverges from just the primary id link that we'd love to be the only one to avoid the confusion of having two differently named anchors for the same section. Is there a way around this?

Simply reverse engineer the section heading to produce the same ID. (This may require some trial and error). Unfortunately, GitHub doesn't let you specify an explicit ID, even though this is a core feature of AsciiDoc.

File: README.adoc (hunk 3 above)
Comment (mojavelinux) [reply README.adoc]: Again, set the `:toc: macro` in the document header, then put `toc::[]` whereever you want the ToC to be placed.

In other words:

```
= Document Title
:toc: macro

yada

yada

toc::[]

yada

yada
```

File: README.adoc (hunk 4 above)
Comment (mojavelinux) [reply README.adoc]: It's up to you. I'm just letting you know that "source" is no longer required. This was an improvement that came in Asciidoctor 2.0.0. See https://github.com/asciidoctor/asciidoctor/issues/1117#issuecomment-475418597

File: README.adoc (hunk 2 above)
Comment (mojavelinux) [reply README.adoc]: I absolutely hate hard wrapping prose at a fixed column. Aside from constantly having to worry about reflow, the main issue is that you make one change at the start of the paragraph and every line in the paragraph gets shifted. This is terrible for collaborative editing. I advocate treating prose the way we do code. Each statement gets its own line. In prose, a statement is a sentence. That way, each change gets scoped to a single line. If you aren't going to use sentence-per-line (aka ventilated prose), the next best thing is soft wrapping.

Here's more context: https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

File: README.adoc [hunk 7]
```diff
@@ -74,12 +58,10 @@ describe Article do
 end
 ----
 
-'''
+=== Empty line between example groups [[empty-lines-between-describes]]
```
Comment (mojavelinux) [inline README.adoc]: If you're going to assign a custom ID to a section, I recommend this syntax:

```
[#empty-lines-between-describes]
=== Empty lines between example groups
```

I also noticed that GitHub now acknowledges explicit IDs everywhere, so if you don't want to rely on the auto-generated ID, just use this syntax to assign a custom one.

**Comment by pirj** [comment]
@mojavelinux Thanks for providing precious information on how to better structure the guide. Followed your recommendations, everything worked like a charm.

I was a bit sceptic regarding moving to AsciiDoc, now I don't have a single doubt that the guide's look and internal representation have improved.

@bbatsov @andyw8 Appreciate if you take a look.

Each recommendation has a title and is represented by an anchorable section.
The old anchors for external resources linking to the guide.
Since the guide had some `_NOTE_`, they became admonitions.

**Comment by mojavelinux** [comment]
@pirj I'm so happy to hear that. I was glad to help you on your journey.

> The old anchors for external resources linking to the guide.

Ah, now I see why some sections have an inline ID. In that case, I agree it's the right way to add a secondary ID for legacy purposes. :+1:

> Since the guide had some _NOTE_, they became admonitions.

That's the beauty of AsciiDoc right there. Semantics FTW!

**Comment by mojavelinux** [comment]
> Ah, now I see why some sections have an inline ID. In that case, I agree it's the right way to add a secondary ID for legacy purposes.

I'll reiterate that I'm particularly excited to learn that GitHub now honors AsciiDoc anchors (e.g., `[[anchor]]`), so you're free to sprinkle them wherever you need them.

**Comment by mojavelinux** [comment]
And btw, it looks really great!

**Comment by bbatsov** [comment]
@pirj Looks good to me too! Only the TOC's heading is a bit weird - its actually smaller than the elements in the TOC. 

Now I guess we'll have to follow suit with the other guides and afterwards we can also publish them on some simple site using git submodules or something like that.

**Comment by mojavelinux** [comment]
> Only the TOC's heading is a bit weird - its actually smaller than the elements in the TOC.

You can fine tune this using the `toc-title` attribute. You can either disable it:

```
:!toc-title:
```

or make it big and bold:

```
:toc-title: pass:[<h3>Table of Contents</h3>]
```

**Comment by pirj** [comment]
Thanks a lot for the help guys!

**Pull request merged**